src/core/cluster/quorum.rs

```diff
@@ -14,99 +16,187 @@ impl Quorum {
+                    debug!("PUT[{}] local node put successful", key);
+                }
+                Err(e) => {
+                    error!("{}", e);
+                }
+            }
+        } else {
+            debug!("skip local put");
+        }
+        debug!("PUT[{}] need at least {} additional puts", key, at_least);
+
+        debug!("PUT[{}] ~~~PUT TO REMOTE NODES~~~", key);
+        let (tasks, mut errors) = self.put_remote_nodes(key, data.clone(), at_least).await;
+        remote_ok_count += at_least - errors.len();
+        let failed_nodes = errors
+            .iter()
+            .map(|e| e.node_name().to_string())
             .collect::<Vec<_>>();
-        let ok_count = total_ops - errors.len();
-
-        debug!("ok: {}/{} quorum: {}", ok_count, total_ops, self.quorum);
-        if ok_count == total_ops {
-            Ok(())
+        let q = self.clone();
+        if remote_ok_count + local_put_ok >= self.quorum {
+            if tasks.is_empty() {
+                Ok(())
+            } else {
+                debug!("PUT[{}] spawn {} background put tasks", key, tasks.len());
+                warn!("PUT[{}]  failed nodes: {:?}", key, failed_nodes);
+                tokio::spawn(q.background_put(tasks, key, data, failed_nodes));
+                Ok(())
+            }
         } else {
-            let mut additional_remote_writes = match ok_count {
-                0 => self.quorum, //@TODO take value from config
-                value if value < self.quorum => 1,
-                _ => 0,
-            };
-
-            let vdisk_id = self.mapper.id_from_key(key);
-            debug!("get names of the failed nodes");
-            let node_names = errors.iter().map(|n| n.node_name().to_owned()).collect();
-            debug!("create operation Backend alien, id: {}", vdisk_id);
-            let operation = Operation::new_alien(vdisk_id);
-            let local_put =
-                put_local_all(&self.backend, node_names, key, data.clone(), operation).await;
-
-            if local_put.is_err() {
-                debug!("local put failed, add another remote node");
-                additional_remote_writes += 1;
+            warn!(
+                "PUT[{}] quorum was not reached. ok {}, quorum {}, errors: {:?}",
+                key,
+                remote_ok_count + local_put_ok,
+                self.quorum,
+                errors
+            );
+            if let Err(err) = self.put_aliens(failed_nodes, key, data).await {
+                if errors.is_empty() {
+                    error!("PUT[{}] smth wrong with cluster/node configuration", key);
+                    error!(
+                        "PUT[{}] local_put_ok: {}, remote_ok_count: {}, quorum: {},no errors",
+                        key, local_put_ok, remote_ok_count, self.quorum
+                    );
+                    Err(err)
+                } else {
+                    let err = errors.remove(errors.len() - 1).into_inner();
```

Comment:
why delete last one?

Reply:
In that case we return last occured error.

Reply:
Can we use `errors.last().into_inner()` ?